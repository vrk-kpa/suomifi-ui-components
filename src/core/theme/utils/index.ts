export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
