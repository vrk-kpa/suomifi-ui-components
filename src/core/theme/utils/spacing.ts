import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (props: SpacingProps | undefined): string => {
  if (!props) return '';

  const cssStyles = Object.entries(props)
    .map(([key, value]) =>
      getCSSSpacing(defaultSuomifiTheme, key as keyof SpacingProps, value),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return `margin-right: ${amount}; margin-left: ${amount};`;
    case 'my':
      return `margin-top: ${amount}; margin-bottom: ${amount};`;
    case 'px':
      return `padding-right: ${amount}; padding-left: ${amount};`;
    case 'py':
      return `padding-top: ${amount}; padding-bottom: ${amount};`;
    default:
      return `${[cssSelector[key]]}: ${amount};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
