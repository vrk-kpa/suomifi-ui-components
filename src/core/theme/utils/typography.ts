import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
