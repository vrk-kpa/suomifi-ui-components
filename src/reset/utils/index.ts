import { css } from 'styled-components';

const normalize = {
  html:
    'line-height: 1.15; -ms-text-size-adjust: 100%; -webkit-text-size-adjust: 100%;',
  h1: 'font-size: 2em; margin: 0.67em 0;',
  'dl dl': 'margin: 0;',
  'dl ol': 'margin: 0;',
  'dl ul': 'margin: 0;',
  'ol dl': 'margin: 0;',
  'ul dl': 'margin: 0;',
  'ol ol': 'margin: 0;',
  'ol ul': 'margin: 0;',
  'ul ol': 'margin: 0;',
  'ul ul': 'margin: 0;',
  hr: 'box-sizing: content-box; height: 0; overflow: visible;',
  main: 'display: block;',
  pre: 'font-family: monospace, monospace; font-size: 1em;',
  a: 'background-color: transparent;',
  'abbr[title]':
    'text-decoration: underline; text-decoration: underline dotted;',
  b: 'font-weight: bolder;',
  strong: 'font-weight: bolder;',
  code: 'font-family: monospace, monospace; font-size: 1em;',
  kbd: 'font-family: monospace, monospace; font-size: 1em;',
  samp: 'font-family: monospace, monospace; font-size: 1em;',
  small: 'font-size: 80%;',
  audio: 'display: inline-block;',
  video: 'display: inline-block;',
  'audio:not([controls])': 'display: none; height: 0;',
  img: 'border-style: none;',
  'svg:not(:root)': 'overflow: hidden;',
  button:
    'margin: 0;overflow: visible; text-transform: none;-webkit-appearance: button;',
  input: 'margin: 0;overflow: visible;',
  select: 'margin: 0;text-transform: none;',
  '[type="button"]': '-webkit-appearance: button;',
  '[type="reset"]': '-webkit-appearance: button;',
  '[type="submit"]': '-webkit-appearance: button;',
  fieldset: 'padding: 0.35em 0.75em 0.625em;',
  legend:
    'box-sizing: border-box; color: inherit; display: table; max-width: 100%; white-space: normal;',
  progress: 'display: inline-block; vertical-align: baseline;',
  textarea: 'margin: 0; overflow: auto;',
  '[type="checkbox"]': 'box-sizing: border-box; padding: 0;',
  '[type="radio"]': 'box-sizing: border-box; padding: 0;',
  '[type="search"]': '-webkit-appearance: textfield; outline-offset: -2px;',
  '::-webkit-inner-spin-button': 'height: auto;',
  '::-webkit-outer-spin-button': 'height: auto;',
  '::-webkit-input-placeholder': 'color: inherit; opacity: 0.54;',
  '::-webkit-search-decoration': '-webkit-appearance: none;',
  '::-webkit-file-upload-button': '-webkit-appearance: button; font: inherit;',
  '::-moz-focus-inner': 'border-style: none; padding: 0;',
  ':-moz-focusring': 'outline: 1px dotted ButtonText;',
  ':-moz-ui-invalid': 'box-shadow: none;',
  details: 'display: block;',
  dialog:
    // eslint-disable-next-line max-len
    'background-color: white; border: solid; color: black; display: block; height: -moz-fit-content; height: -webkit-fit-content; height: fit-content; left: 0; margin: auto; padding: 1em; position: absolute; right: 0; width: -moz-fit-content; width: -webkit-fit-content; width: fit-content;',
  'dialog:not([open])': 'display: none;',
  summary: 'display: list-item;',
  canvas: 'display: inline-block;',
  template: 'display: none;',
  '[hidden]': 'display: none;',
} as Record<string, string>;

const common = css`
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;

  ::before,
  ::after {
    box-sizing: border-box;
  }
`;

const resets = {
  normalize,
  common,
};

const selectorDeclaration = (selector: string, styles: string) =>
  `&${selector} {
    ${styles}
  }`;

const selectorsAndCss = (selector: string) => {
  const styles = !!normalize && normalize[selector];
  return !!styles ? selectorDeclaration(selector, styles) : '';
};

/**
 * Return selectors with resets in Emotion-css
 * @param selectors string or array of strings
 */
const resetWithSelectors = (selectors: string | string[]): string => {
  if (Array.isArray(selectors) && selectors.length > 0) {
    return `
      ${selectors.map((s) => selectorsAndCss(s)).join('')}
    `;
  }
  if (typeof selectors === 'string') {
    return selectorsAndCss(selectors);
  }
  return '';
};

export { resets, resetWithSelectors };

// TODO: Add tests
// TODO: create versioned interface for normalize.css
