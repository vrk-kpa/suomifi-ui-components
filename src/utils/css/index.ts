export { alphaHex, hslaToHex } from './colors';
export { allStates, absolute } from './pseudo';
export {
  clearfix,
  cssObjectToCss,
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from './utils';
