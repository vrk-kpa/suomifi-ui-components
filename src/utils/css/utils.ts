import { css, FlattenSimpleInterpolation } from 'styled-components';
import { ValueUnit } from 'suomifi-design-tokens';
export { FlattenSimpleInterpolation };

export const clearfix = css`
  &:after {
    display: block;
    content: '';
    clear: both;
  }
`;

/**
 * Return FlattenSimpleInterpolation compatible string or number
 * @param cssValue compatible value or {value, unit} pair-object
 */
const cssValueToString = (cssValue: number | string | ValueAndUnit) => {
  if (!!cssValue && typeof cssValue === 'object' && 'value' in cssValue) {
    const { value, unit } = cssValue;
    return !!unit ? `${value}${unit}` : value;
  }
  return cssValue;
};

const camelToSnake = (string: string) =>
  string.replace(/[\w]([A-Z])/g, m => `${m[0]}-${m[1]}`).toLowerCase();
/**
 * Convert CSSObject to CSS FlattenSimpleInterpolation
 * @param value CSSObject
 */
export const cssObjectToCss = <T>(value: T) => css`
  ${Object.entries(value)
    .map(([key, value]) => `${camelToSnake(key)}: ${cssValueToString(value)};`)
    .join('')}
`;

/**
 * Convert CSSObject tokens to CSS FlattenSimpleInterpolation
 * @param tokens Tokens of CSSObjects
 */
export const cssObjectsToCss = <T, K extends keyof T>(tokens: T) =>
  Object.entries(tokens).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: cssObjectToCss(value),
    }),
    {} as { [key in K]: FlattenSimpleInterpolation },
  );
