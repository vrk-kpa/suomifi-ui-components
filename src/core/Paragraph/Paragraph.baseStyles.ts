import { css } from 'styled-components';
import { SuomifiTheme } from '../theme';
import { element, font } from '../theme/reset';
import { MarginProps, buildSpacingCSS } from '../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${element(theme)}
  ${font(theme)('bodyText')}
  margin: 0;
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
`;
