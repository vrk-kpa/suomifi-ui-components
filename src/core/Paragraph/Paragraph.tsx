import React, { forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../theme/utils/spacing';
import {
  ColorProp,
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../theme';
import { baseStyles } from './Paragraph.baseStyles';
import { HtmlP, HtmlPProps } from '../../reset/HtmlP/HtmlP';
import { filterDuplicateKeys } from '../../utils/common/common';

const baseClassName = 'fi-paragraph';

export interface ParagraphProps extends HtmlPProps, MarginProps {
  /** Sets a color for the text. Colors from SuomifiTheme are available */
  color?: ColorProp;
  /** Ref object is forwarded to the paragraph element. Alternative to React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLParagraphElement>;
}

const StyledParagraph = styled(
  ({
    className,
    theme,
    style,
    globalMargins,
    ...rest
  }: ParagraphProps & SuomifiThemeProp & GlobalMarginProps) => {
    const [_marginProps, passProps] = separateMarginProps(rest);

    return (
      <HtmlP
        className={classnames(baseClassName, className)}
        {...passProps}
        style={style}
      />
    );
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.paragraph,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const Paragraph = forwardRef<HTMLParagraphElement, ParagraphProps>(
  (props, ref) => (
    <SpacingConsumer>
      {({ margins }) => (
        <SuomifiThemeConsumer>
          {({ suomifiTheme }) => (
            <StyledParagraph
              theme={suomifiTheme}
              globalMargins={margins}
              forwardedRef={ref}
              {...props}
            />
          )}
        </SuomifiThemeConsumer>
      )}
    </SpacingConsumer>
  ),
);

Paragraph.displayName = 'Paragraph';
export { Paragraph };
