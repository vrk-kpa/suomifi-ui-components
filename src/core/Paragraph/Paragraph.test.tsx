import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../utils/test';

import { Paragraph } from './Paragraph';

const TestParagraph = (
  <div data-testid="test-paragraph">
    <Paragraph>Hey this is test</Paragraph>
  </div>
);

describe('margin', () => {
  it('should have margin style from margin prop', () => {
    const { container } = render(<Paragraph margin="xs">Test</Paragraph>);
    expect(container.firstChild).toHaveStyle('margin: 10px');
  });

  it('should have margin prop overwritten from style prop', () => {
    const { container } = render(
      <Paragraph margin="xs" style={{ margin: 2 }}>
        Test
      </Paragraph>,
    );
    expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
  });
});

test('calling render with the same component on the same container does not remount', () => {
  const ParagraphRendered = render(TestParagraph);
  const { container } = ParagraphRendered;
  expect(container.firstChild).toMatchSnapshot();
});

test('should not have basic accessibility issues', axeTest(TestParagraph));
