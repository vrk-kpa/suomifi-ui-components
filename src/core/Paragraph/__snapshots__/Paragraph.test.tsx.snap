// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c1 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: hsl(0,0%,16%);
}

<div
  data-testid="test-paragraph"
>
  <p
    class="fi-paragraph c0 c1"
  >
    Hey this is test
  </p>
</div>
`;
